src/plugins/data/public/search/search_service.ts

```diff
@@ -92,22 +95,28 @@ export class SearchService implements Plugin<ISearchSetup, ISearchStart> {
 
   public setup(
     core: CoreSetup,
-    { packageInfo, query, getInternalStartServices }: SearchServiceSetupDependencies
+    { expressions, packageInfo, query, getInternalStartServices }: SearchServiceSetupDependencies
   ): ISearchSetup {
     this.esClient = getEsClient(core.injectedMetadata, core.http, packageInfo);
     this.registerSearchStrategyProvider(SYNC_SEARCH_STRATEGY, syncSearchStrategyProvider);
     this.registerSearchStrategyProvider(ES_SEARCH_STRATEGY, esSearchStrategyProvider);
 
     const aggTypesSetup = this.aggTypesRegistry.setup();
+
+    // register each agg type
     const aggTypes = getAggTypes({
       query,
       uiSettings: core.uiSettings,
       getInternalStartServices,
     });
-
     aggTypes.buckets.forEach(b => aggTypesSetup.registerBucket(b));
     aggTypes.metrics.forEach(m => aggTypesSetup.registerMetric(m));
 
+    // register expression functions for each agg type
+    const aggFunctions = getAggTypesFunctions();
```

Comment:
I see you tried to do it in a way like we do for `aggTypes`.  But maybe we can create a factory method (or you can use some other pattern) to encapsulate forEach methods and make the setup hook easier for reading. It will be awesome if you will do it for `aggTypes` and `aggFuncitons`

Reply:
@alexwizp Just to make sure I follow; how would you envision this looking? Just removing the `forEach` from `search_service` entirely & passing `expressions` down so it is handled inside of the function instead?

Reply:
The reason I ask is because readability is what I was aiming for here, but of course I'm always open to ideas to make things more readable. Rather than having the magic of stuff getting registered in another function, I liked the idea of seeing very clearly what this service was registering during `setup`.

Reply:
As a start I removed the bucket/metric differentiation per Peter's suggestion; lmk what your thoughts are and I'm happy to adjust further